File: llm_harness/clients/parser.py
"""Response parsing utilities with metadata extraction."""

from __future__ import annotations

from collections.abc import Generator

from langchain_core.messages import AIMessage
from pydantic import BaseModel, ConfigDict


class StructuredOutput(BaseModel):
    """Parse the raw message from the structured output response.

    >>> StructuredOutput = ChatOpenAI.with_structured_output(Schema, include_raw=True)
    """

    model_config = ConfigDict(extra="ignore")
    raw: AIMessage
    parsed: BaseModel


def _to_int(value: object) -> int:
    return int(value or 0)


def _to_float(value: object) -> float:
    return float(value or 0.0)


def get_metadata(ai_message: AIMessage) -> tuple[int, int, float]:
    """Return (input_tokens, output_tokens, cost) from an AIMessage.

    - Prefer `AIMessage.usage_metadata` (LangChain standard; works for Gemini).
    - Fall back to `response_metadata["token_usage"]` (OpenAI/OpenRouter; includes cost here).
    - Fall back to legacy Gemini shapes under `response_metadata["usage_metadata"]`.

    Cost may be unavailable for some providers; returns 0.0.
    """
    usage_metadata = getattr(ai_message, "usage_metadata", None)
    if isinstance(usage_metadata, dict) and usage_metadata:
        return (
            _to_int(usage_metadata.get("input_tokens")),
            _to_int(usage_metadata.get("output_tokens")),
            0.0,
        )

    response_metadata = getattr(ai_message, "response_metadata", None)
    if not isinstance(response_metadata, dict) or not response_metadata:
        return 0, 0, 0.0

    token_usage = response_metadata.get("token_usage")
    if isinstance(token_usage, dict) and token_usage:
        return (
            _to_int(token_usage.get("prompt_tokens")),
            _to_int(token_usage.get("completion_tokens")),
            _to_float(token_usage.get("cost")),
        )

    legacy_usage = response_metadata.get("usage_metadata")
    if isinstance(legacy_usage, dict) and legacy_usage:
        input_tokens = _to_int(legacy_usage.get("prompt_token_count") or legacy_usage.get("input_token_count"))
        output_tokens = _to_int(legacy_usage.get("candidates_token_count") or legacy_usage.get("output_token_count"))
        return input_tokens, output_tokens, 0.0

    return 0, 0, 0.0


def _extract_reasoning(content_blocks: list[dict]) -> str | None:
    """Extract reasoning from response content_blocks."""
    if not content_blocks:
        return None

    first_block = content_blocks[0]
    if reasoning := first_block.get("reasoning"):
        return reasoning

    # Check for nested content structure (e.g. from some providers)
    if (
        (extras := first_block.get("extras"))
        and isinstance(extras, dict)
        and (nested_content := extras.get("content"))
        and isinstance(nested_content, list)
        and nested_content
        and isinstance(nested_content[-1], dict)
    ):
        return nested_content[-1].get("text")

    return None


def parse_invoke(
    response: AIMessage,
    include_reasoning: bool = False,
) -> str | tuple[str | None, str]:
    """Parse response to extract answer and optionally reasoning."""
    answer = response.content_blocks[-1]["text"]
    if include_reasoning:
        reasoning = _extract_reasoning(response.content_blocks)
        return reasoning, answer
    return answer


def parse_batch(
    responses: list[AIMessage],
    include_reasoning: bool = False,
) -> list[str] | list[tuple[str | None, str]]:
    """Parse batched responses, optionally with reasoning."""
    return [parse_invoke(response, include_reasoning) for response in responses]


def get_stream_generator(
    stream: Generator[AIMessage],
    include_reasoning: bool = False,
) -> Generator[str | tuple[str, str | None]]:
    """Yield streaming chunks, optionally with reasoning."""
    reasoning_yielded = False

    for chunk in stream:
        if not (blocks := getattr(chunk, "content_blocks", None)):
            continue

        if include_reasoning and not reasoning_yielded and (reasoning := _extract_reasoning(blocks)):
            reasoning_yielded = True
            yield (reasoning, None)

        if answer := blocks[-1].get("text"):
            yield (None, answer) if include_reasoning else answer


def parse_stream(
    stream: Generator[AIMessage],
    include_reasoning: bool = False,
) -> str | tuple[str | None, str]:
    """Print streamed chunks and return the final result."""
    reasoning = None
    answer_parts: list[str] = []

    for item in get_stream_generator(stream, include_reasoning):
        if isinstance(item, tuple):
            reasoning_chunk, answer_chunk = item
            if reasoning_chunk is not None:
                reasoning = reasoning_chunk
                print(f"Reasoning: {reasoning}", flush=True)
            if answer_chunk is not None:
                answer_parts.append(answer_chunk)
                print(answer_chunk, end="", flush=True)
        else:
            answer_parts.append(item)
            print(item, end="", flush=True)

    answer = "".join(answer_parts)
    return (reasoning, answer) if include_reasoning else answer

File: llm-harness-js/src/clients/parser.ts
export interface StructuredOutput<TRaw = unknown, TParsed = unknown> {
  raw: TRaw;
  parsed: TParsed;
}

function toInt(value: unknown): number {
  return Number(value ?? 0);
}

function toFloat(value: unknown): number {
  return Number(value ?? 0);
}

export function getMetadata(
  aiMessage: Record<string, unknown>,
): [number, number, number] {
  const usageMetadata = aiMessage.usage_metadata;
  if (usageMetadata && typeof usageMetadata === "object") {
    const typed = usageMetadata as Record<string, unknown>;
    return [toInt(typed.input_tokens), toInt(typed.output_tokens), 0];
  }

  const responseMetadata = aiMessage.response_metadata;
  if (!responseMetadata || typeof responseMetadata !== "object") {
    return [0, 0, 0];
  }

  const response = responseMetadata as Record<string, unknown>;
  const tokenUsage = response.token_usage;
  if (tokenUsage && typeof tokenUsage === "object") {
    const typed = tokenUsage as Record<string, unknown>;
    return [
      toInt(typed.prompt_tokens),
      toInt(typed.completion_tokens),
      toFloat(typed.cost),
    ];
  }

  const legacyUsage = response.usage_metadata;
  if (legacyUsage && typeof legacyUsage === "object") {
    const typed = legacyUsage as Record<string, unknown>;
    const inputTokens = toInt(
      typed.prompt_token_count ?? typed.input_token_count,
    );
    const outputTokens = toInt(
      typed.candidates_token_count ?? typed.output_token_count,
    );
    return [inputTokens, outputTokens, 0];
  }

  return [0, 0, 0];
}

function extractReasoning(
  contentBlocks: Record<string, unknown>[],
): string | null {
  const firstBlock = contentBlocks[0];
  if (!firstBlock) {
    return null;
  }

  const reasoning = firstBlock.reasoning;
  if (typeof reasoning === "string") {
    return reasoning;
  }

  const extras = firstBlock.extras;
  if (!extras || typeof extras !== "object") {
    return null;
  }

  const nestedContent = (extras as Record<string, unknown>).content;
  if (!Array.isArray(nestedContent) || nestedContent.length === 0) {
    return null;
  }

  const last = nestedContent.at(-1);
  if (!last || typeof last !== "object") {
    return null;
  }

  const text = (last as Record<string, unknown>).text;
  return typeof text === "string" ? text : null;
}

function extractAnswer(response: Record<string, unknown>): string {
  const blocks = response.content_blocks;
  if (!Array.isArray(blocks) || blocks.length === 0) {
    return "";
  }
  const last = blocks.at(-1) as Record<string, unknown>;
  return typeof last.text === "string" ? last.text : "";
}

export function parseInvoke(
  response: Record<string, unknown>,
  includeReasoning = false,
): string | [string | null, string] {
  const answer = extractAnswer(response);
  if (!includeReasoning) {
    return answer;
  }

  const blocks = Array.isArray(response.content_blocks)
    ? (response.content_blocks as Record<string, unknown>[])
    : [];
  const reasoning = extractReasoning(blocks);
  return [reasoning, answer];
}

export function parseBatch(
  responses: Record<string, unknown>[],
  includeReasoning = false,
): Array<string | [string | null, string]> {
  return responses.map((response) => parseInvoke(response, includeReasoning));
}

export async function* getStreamGenerator(
  stream: AsyncIterable<Record<string, unknown>>,
  includeReasoning = false,
): AsyncGenerator<string | [string | null, string | null]> {
  let reasoningYielded = false;

  for await (const chunk of stream) {
    const blocks = chunk.content_blocks;
    if (!Array.isArray(blocks) || blocks.length === 0) {
      continue;
    }

    const typedBlocks = blocks as Record<string, unknown>[];

    if (includeReasoning && !reasoningYielded) {
      const reasoning = extractReasoning(typedBlocks);
      if (reasoning) {
        reasoningYielded = true;
        yield [reasoning, null];
      }
    }

    const last = typedBlocks.at(-1);
    const answer = typeof last?.text === "string" ? last.text : null;
    if (answer) {
      yield includeReasoning ? [null, answer] : answer;
    }
  }
}

export async function parseStream(
  stream: AsyncIterable<Record<string, unknown>>,
  includeReasoning = false,
): Promise<string | [string | null, string]> {
  let reasoning: string | null = null;
  const answerParts: string[] = [];

  for await (const item of getStreamGenerator(stream, includeReasoning)) {
    if (Array.isArray(item)) {
      const [reasoningChunk, answerChunk] = item;
      if (reasoningChunk !== null) {
        reasoning = reasoningChunk;
        // Keep parity with Python parser side effects during stream parsing.
        // eslint-disable-next-line no-console
        console.log(`Reasoning: ${reasoning}`);
      }
      if (answerChunk !== null) {
        answerParts.push(answerChunk);
        // eslint-disable-next-line no-console
        process.stdout.write(answerChunk);
      }
    } else {
      answerParts.push(item);
      // eslint-disable-next-line no-console
      process.stdout.write(item);
    }
  }

  const answer = answerParts.join("");
  return includeReasoning ? [reasoning, answer] : answer;
}
